src/ui.c

```diff
@@ -186,3 +201,94 @@ void ui_render()
 
 	SDL_RenderPresent(renderer);
 }
+
+/* Translate real input keys to MS keyboard matrix
+ *
+ * The lookup matrix is currently [10][8], we just walk the whole thing like
+ * one long buffer and do the math later to figure out exactly what bit was
+ * pressed.
+ *
+ * TODO: Would it make sense to rework keyTranslateTable to a single buffer
+ * anyway? Right now the declaration looks crowded and would need some rework
+ * already.
+ */
+static void ui_set_ms_kbd(ms_ctx* ms, int scancode, int eventtype)
+{
+	uint32_t i = 0;
+	int32_t *keytbl_ptr = &sdl_to_ms_kbd_LUT[0][0];
+
+	for (i = 0; i < (sizeof(sdl_to_ms_kbd_LUT)/sizeof(int32_t)); i++) {
+		if (scancode == *(keytbl_ptr + i)) {
+			/* Couldn't avoid the magic numbers below. As noted,
+			 * kTT array is [10][8], directly mapping the MS matrix
+			 * of 10 bytes to represent the whole keyboard. Divide
+			 * by 8 to get the uint8_t the scancode falls in, and mod
+			 * 8 to get the bit in that uint8_t that matches the code.
+			 */
+			if (eventtype == SDL_KEYDOWN) {
+				ms->key_matrix[i/8] &= ~((uint8_t)1 << (i%8));
+				break;
+			} else {
+				ms->key_matrix[i/8] |= ((uint8_t)1 << (i%8));
+				break;
+			}
+		}
+	}
+}
+
+int ui_kbd_process(ms_ctx *ms)
+{
+
+	SDL_Event event;
+	// Check SDL events
+	while (SDL_PollEvent(&event))
+	{
+		/* Exit if SDL quits, or Escape key was pushed */
+		if ((event.type == SDL_QUIT) ||
+		  ((event.type == SDL_KEYDOWN) &&
+			(event.key.keysym.sym == SDLK_ESCAPE))) {
+			return 1;
+		}
+
+
+		/* Handle other input events */
+		if ((event.type == SDL_KEYDOWN) ||
+		  (event.type == SDL_KEYUP)) {
+
+			/* First, check to see if F12 was pressed */
+			if (event.key.keysym.sym == SDLK_F12) {
+				if (event.type == SDL_KEYDOWN) {
+					ms->power_button = 1;
+					if (ms->power_state == MS_POWERSTATE_OFF) {
+						printf("POWER ON\n");
+						ms_power_on_reset(ms);
+						ui_splashscreen_hide();
+					}
+				} else {
+					ms->power_button = 0;
+				}
+			}
+
+			/* Keys pressed whie right ctrl is held */
```

Comment:
[nit] spelling: `whie`

Reply:
I'll make a note of it and clean it up next time I touch ui.c